### DISCUSSION TITLE
EV3 Control Non-motor Device

### DISCUSSION THREAD

**Discussion by shoringit** [pybricks/support#1168] [Q&A] [answered]
I'm making the transition from EV3-g over to MicroPython for a school robotics team I coach.  I'm currently in the "figure it out" stage, where I'm trying to get everything we had working to work in the new language and prove it out before the upcoming season starts.  I have some experience with Python, but not much (most of my programming is in C, C++, Java and web development (HTML, CSS, PHP, JavaScript, PHP, etc.).  I'll also note that I have a hate-hate relationship with Linux - it seems to hate me and I hate it back...

My current stumbling block surrounds a non-lego device hooked to a motor port.  We used the Motor+ and Motor- as outputs to drive this device.  In the EV3 world, I've used both a custom block from MindSensors (for their [relay driver](http://www.mindsensors.com/ev3-and-nxt/82-relay-driver-for-nxt-or-ev3)) as well as the unregulated motor block.  Setting +100 engages one "output", and -100 does the other one.

In MicroPython, I've tried to use the DCMotor function to do this (in iodevices).  Here's the code in question:

`lightControl = DCMotor(Port.D)`

But when I try and run it, I get this:

`Traceback (most recent call last):
  File "/home/robot/test/main.py", line 38, in <module>
OSError: [Errno 5] EIO: 

Unexpected hardware input/output error with a motor or sensor:
--> Try unplugging the sensor or motor and plug it back in again.
--> To see which sensor or motor is causing the problem,
    check the line in your script that matches
    the line number given in the 'Traceback' above.
--> Try rebooting the hub/brick if the problem persists.`

Since the DCMotor has no real feedback, I'm surprised of the error.  I guess it does a resistance check, etc.  So, the question is how can I control this or disable the check?

I'm sure one way would be to drive the outputs directly.  I came across [this](https://github.com/pybricks/support/issues/251#issuecomment-774066191), for driving the built-in LEDs directly, but it's beyond my knowledge to convert this to driving a port.  And I'm hoping that there's a better way (that would be easier for kids to understand)...  Any help would be appreciated.

Thanks,

Danny

**Comment by dlech** [comment]
By default the input and output ports drivers on ev3dev are set to automatically detect what is connected to the port. Some devices, like the one you are using, can't be automatically detected. So you have to tell the port that is connected to it. 

You can do this manually using the Brickman interface on the EV3 by going to the *Device Browser*, selecting *Ports* then select port D and then select *Set mode* and change the mode to `dc-motor`.

You can also do this programatically like this:

```python
...

# Place port D in `dc-motor` mode.
with open("/sys/class/lego-port/port7/mode", "w") as f:
    f.write("dc-motor")

# Probably need to wait for the `dc-motor` driver to register.
wait(500)

light_control = DCMotor(Port.D)

...
```

**Reply by shoringit** [reply]
I looked on the interface, and it was set to DC-Motor.  I tried setting it to the other options, and it always gets reset to DC-Motor when I run my code.

I tried adding your code, and it didn't make a difference - same error.  Then, I noticed in the Brickman interface for that port that the device name was "port7", so I changed your code from port6 to port7 and it didn't make any difference...

Any other ideas?

**Reply by dlech** [reply]
Manually setting the port mode to `dc-motor` worked for me. I guess I have one port disabled for debugging on my EV3 which is why I had the off-by-one error on the port number.

Did you try waiting longer after changing the port mode?

**Reply by shoringit** [reply]
Yes, I increased it to 10000 with no difference - same error.  And I double-checked that the port is set to DC-Motor in Brickman, so I don't think that's the problem.  But I don't know enough about this yet to know what the problem might be...

**Reply by dlech** [reply]
After you change the port mode to `dc-motor` under *Ports* in the *Device Browser* is there a motor listed under *Motors* in the *Device Browser*? For example, on my EV3, I see `rcx-motor at ev3-ports:outD`.

**Reply by shoringit** [reply]
Yes, it shows "rcx-motor at ev3-ports:outD"

**Reply by dlech** [reply]
Once it shows this, the line of code `light_control = DCMotor(Port.D)` is working for me.

**Reply by shoringit** [reply]
Maybe you're running a different version than I am?...  I downloaded my image from [here](https://www.ev3dev.org/downloads/).  I'm assuming it contains v2, but I don't even know how to check that...

**Reply by dlech** [reply]
I'm glad you asked. I was using a different version, but just checked it again with Pybricks v2.

```
robot@pybricks-ev3:~$ pybricks-repl
Pybricks MicroPython v1.11 on 2020-05-06; linux version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> from pybricks.iodevices import DCMotor
>>> m = DCMotor(Port.D)
>>> 
```

**Reply by shoringit** [reply]
So, mine shows v1.11 like yours, but mind has an April date, not May...  I'm betting that's what's going on.  Now to figure out why mine is different...

```

robot@pybricks-ev3:~$ pybricks-repl
Pybricks MicroPython v1.11 on 2020-04-10; linux version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> from pybricks.iodevices import DCMotor
>>> m = DCMotor(Port.D)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 5] EIO: 

Unexpected hardware input/output error with a motor or sensor:
--> Try unplugging the sensor or motor and plug it back in again.
--> To see which sensor or motor is causing the problem,
    check the line in your script that matches
    the line number given in the 'Traceback' above.
--> Try rebooting the hub/brick if the problem persists.

```

**Reply by dlech** [reply]
To update, ensure your EV3 is connected to the Internet and run:

```
sudo apt udpate
sudo apt upgrade
```

**Reply by shoringit** [reply]
I haven't hooked it to the net yet - probably will soon as I'm about to start experimenting with Bluetooth communication, etc. next.  So, I just grabbed a different image, and that did the trick.  Thanks for the help.

**Comment by shoringit** [comment, marked as answer]
So, I grabbed a different image than I had used prevously ([this one](https://education.lego.com/en-us/product-resources/mindstorms-ev3/teacher-resources/python-for-ev3) now) and things work as they should.  I have the same date code when running `pybricks-repl` and more importantly, the code works - I can control my device as I did under EV3-g.

Thanks for the help!!

**Reply by d12nxt** [reply]
Hello! I'm having the same problem with the DC motor not connecting to the EV3 brick in micropython, I did everything that was said by you and dlech. But none of your and dlech solutions have worked, i have de 2020-05-09 update installed, I've tried the commands in SSH terminal, connecting the cables manually and still nothing. I have 3x5292 motors and 2xRCX 443362 motors that I tried to connect to the EV3 brick with the LEGO original 9V to NXT converter cable and none of them work with your solutions. I know I have an account created today but my old account got stolen.
I'm waiting for your response. Thanks!
Best regards,
Daniel

**Reply by dlech** [reply]
Non-EV3 motors can still be used, they just won't be automatically detected. Can you share the full script that you tried?

**Reply by d12nxt** [reply]
This is the code that I've tried with your methods. As a side note, the speaker beep is for a small debugging. Also, there are other libraries, because this is a test code and after I figure out how to use 9v motors I will use it in the main code.

```python
#!/usr/bin/env pybricks-micropython
from pybricks.hubs import EV3Brick
from pybricks.ev3devices import (Motor, TouchSensor, ColorSensor,
                                 InfraredSensor, UltrasonicSensor, GyroSensor)
from pybricks.parameters import Port, Stop, Direction, Button, Color
from pybricks.tools import wait, StopWatch, DataLog
from pybricks.robotics import DriveBase
from pybricks.media.ev3dev import SoundFile, ImageFile
from pybricks.iodevices import DCMotor

wait(500)

E_Motor = DCMotor(Port.D)
ev3.speaker.beep()
dc(50)
```

**Reply by laurensvalk** [reply]
When the program ends, the motors will stop, so nothing appears to happen.

Try adding something after the `dc(50)`, such as:

```python
E_Motor.dc(50)
wait(10000)
```

**Reply by d12nxt** [reply]
I've tried, but still the code gives the error 5 for not reading the DCMotor library. And I have 6 EV3 bricks and none of them want to communicate with the DCMotors that I have.

**Reply by BertLindeman** [reply]
No idea if this helps, but I ran this program without error.
Some comments in the program, annotated with "/BL".

```python
#!/usr/bin/env pybricks-micropython

# removed some of the imports (not used)

from pybricks.hubs import EV3Brick
#from pybricks.ev3devices import (Motor, TouchSensor, ColorSensor, InfraredSensor, UltrasonicSensor, GyroSensor)
from pybricks.parameters import Port, Stop, Direction, Button, Color
from pybricks.tools import wait, StopWatch, DataLog
#from pybricks.robotics import DriveBase
#from pybricks.media.ev3dev import SoundFile, ImageFile
from pybricks.iodevices import DCMotor

wait(500)
ev3 = EV3Brick()  # /BL added ev3 was not defined
E_Motor = DCMotor(Port.B)
ev3.speaker.beep()
E_Motor.dc(50)  # /BL added "E_Motor." as device
ev3.speaker.beep()
wait(10000)  # /BL added
E_Motor.stop()  # /BL added

```

**Reply by d12nxt** [reply]
It works with your version of the code.
Sorry to everyone that I had to bring up again this discussion, but now it works.
I thank everyone who tried to help me.

**Reply by d12nxt** [reply]
I know it was kind of sudden that it works now, but that's what i wanted to know the structures to make the motors work. Beacuse i will use the structures in a state machine made with micropython.
Thanl you all for helping!

**Reply by laurensvalk** [reply]
Thanks Bert! I hadn't noticed the missing object :-)

Glad it works @d12nxt!

**Comment by dlech** [comment]
> I've tried, but still the code gives the error 5 for not reading the DCMotor library.

Which version of Pybricks MicroPython are you using?

**Reply by d12nxt** [reply]
On VS Code i'm using 2.0.0 and on the brick i have V1.1 with 2020-05-06 update beacuse i verified with pybricks-repl commans in SSH Terminal.
